### PR TITLE
WIP: add initial support for BELs chains

### PR DESCRIPTION
**Pull request by kowalewskijan**
This is an initial version of BELs chains support. This is related to SymbiFlow/nextpnr#262
The main target for now is to get carry chains support for xilinx architectures, but the solution should be applicable for any architecture in the end. The PR includes what type of data we should pass to nextpnr in order to get the constraining step done of the BELs chains.

Signed-off-by: Jan Kowalewski <jkowalewski@antmicro.com>

### CODE REVIEW

File: fpga_interchange/nextpnr.py [hunk 1]
```diff
@@ -81,5 +81,12 @@ def pop(self):
 
     def check_labels(self):
         refs_and_labels = self.refs & self.labels
-        assert len(refs_and_labels) == len(self.refs)
-        assert len(refs_and_labels) == len(self.labels)
+        for ref in refs_and_labels:
+            if ref not in self.refs:
```
Comment (gatecat) [inline fpga_interchange/nextpnr.py:85]: I think this can be simplified to`assert ref in self.refs, ref`?

File: fpga_interchange/chip_info.py [hunk 2]
```diff
@@ -632,6 +632,149 @@ def append_bba(self, bba, label_prefix):
...
+class ChainCoordConfig():
+    def __init__(self, coord, step):
+        self.coord = coord
+        self.step = step
+
+    def append_children_bba(self, bba, label_prefix):
+        pass
+
+    def append_bba(self, bba, label_prefix):
+        bba.str(self.coord)
```
Comment (gatecat) [inline fpga_interchange/chip_info.py:657]: this should definitely be an enum, not an string, similar to my comment on the schema

File: fpga_interchange/chip_info.py [hunk 3]
```diff
@@ -632,6 +632,149 @@ def append_bba(self, bba, label_prefix):
...
+class ChainPattern():
+    def __init__(self, source, sink):
+        self.source = source
+        self.sink = sink
+
+    def append_children_bba(self, bba, label_prefix):
+        pass
+
+    def append_bba(self, bba, label_prefix):
+        bba.str(self.source)
```
Comment (gatecat) [inline fpga_interchange/chip_info.py:644]: constids (`str_id`) not raw strings should be being used for stuff like this - have a look at similar places where bel/cell relationships are being described like `LutBel`

Comment (kowalewskijan) [reply fpga_interchange/chip_info.py]: done, thanks

**Comment by acomodi** [comment]
Superseeded by https://github.com/SymbiFlow/python-fpga-interchange/pull/91, closing

**Pull request closed without merging**